ballerina/client.bal

```diff
@@ -0,0 +1,163 @@
+            method: "tools/list"
+        };
+        JSONRPCServerMessage|stream<JSONRPCServerMessage, error?>|() response = check self.sendRequest(listToolRequest);
+        ListToolsResult listToolResult = check handleListToolResult(response);
+        return listToolResult;
+    }
+
+    public function callTool(CallToolParams params) returns JSONRPCServerMessage|stream<JSONRPCServerMessage, error?>|error {
+        CallToolRequest callToolRequest = {
+            method: "tools/call",
+            params: params
+        };
+        JSONRPCServerMessage|stream<JSONRPCServerMessage, error?>|() response = check self.sendRequest(callToolRequest);
+        if response is () {
+            return error CallToolError("Received invalid response for tool call");
+        }
+        return response;
+    }
+
+    public function waitForCompletion() returns error? {
+        StreamableHttpClientTransport? transport = self.transport;
+        if transport is () {
+            return error UninitializedTransportError("Transport is not initialized for sending requests");
+        }
+
+        return transport.waitForCompletion();
+    }
+
+    public function setNotificationHandler(function (JSONRPCServerMessage) returns error? serverMessageHandler) returns error? {
+        self.serverMessageHandler = serverMessageHandler;
+    }
+
+    private function sendRequest(Request request) returns JSONRPCServerMessage|stream<JSONRPCServerMessage, error?>|error? {
+        StreamableHttpClientTransport? transport = self.transport;
+        if transport is () {
+            return error UninitializedTransportError("Transport is not initialized for sending requests");
+        }
+
+        self.requestMessageId += 1;
+        JSONRPCRequest jsonrpcRequest = {
+            ...request,
+            jsonrpc: JSONRPC_VERSION,
+            id: self.requestMessageId
+        };
+
+        return transport.send(jsonrpcRequest);
+    }
+
+    private function sendNotification(Notification notification) returns error? {
+        StreamableHttpClientTransport? transport = self.transport;
+        if transport is () {
+            return error UninitializedTransportError("Transport is not initialized for sending requests");
+        }
+
+        JSONRPCNotification jsonrpcNotification = {
+            ...notification,
+            jsonrpc: JSONRPC_VERSION
+        };
+
+        _ = check transport.send(jsonrpcNotification);
+
+    }
+
+    private function handleSseStream(stream<JSONRPCServerMessage, error?> 'stream) returns future<error?> {
+        worker SseWorker returns error? {
+            check from JSONRPCServerMessage serverMessage in 'stream
+                do {
+                    ServerMessageHandler? handler = self.serverMessageHandler;
+                    if handler is ServerMessageHandler {
+                        check handler(serverMessage);
+                    }
+                };
+        }
+        return SseWorker;
+    }
```

Comment:
Instead of return a worker we could use async function call here: https://ballerina.io/learn/by-example/asynchronous-function-calls/